File: iq/5/shiftzeros/solution.go
// Copyright 2026 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package shiftzeros

import "iter"

func Shift(nn []int) {
	offset := find(nn, isZero)
	zi := offset
	for ni := range findAll(nn[offset:], isNonZero) {
		swap(nn, zi, offset+ni)
		zi += 1
	}
}

func find(nn []int, f func(n int) bool) int {
	for i, n := range nn {
		if f(n) {
			return i
		}
	}
	return len(nn)
}

func findAll(nn []int, f func(n int) bool) iter.Seq[int] {
	return func(yield func(int) bool) {
		for i, n := range nn {
			if !f(n) {
				continue
			}
			if !yield(i) {
				return
			}
		}
	}
}

func isZero(n int) bool {
	return n == 0
}

func isNonZero(n int) bool {
	return n != 0
}

func swap(nn []int, i, j int) {
	if i == j {
		return
	}
	nn[i], nn[j] = nn[j], nn[i]
}

File: iq/5/shiftzeros/solution.cc
// Copyright 2026 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

//
// NOLINTNEXTLINE
//go:build ignore

#include "iq/5/shiftzeros/solution.h"

#include <cstddef>
#include <utility>
#include <vector>

namespace iq::shiftzeros {
namespace {

constexpr int kZero = 0;

std::size_t indexZero(std::vector<int> nn) {
  std::size_t i = 0;
  while (i < nn.size() && nn[i] != kZero) {
    ++i;
  }
  return i;
}

std::size_t indexNonZero(std::vector<int> nn, std::size_t offset) {
  std::size_t i = offset;
  while (i < nn.size() && nn[i] == kZero) {
    ++i;
  }
  return i;
}

}  // namespace

void Shift(std::vector<int>& nn) {
  std::size_t zero_idx = indexZero(nn);
  if (zero_idx == nn.size()) {
    return;
  }
  for (std::size_t non_zero_idx = indexNonZero(nn, zero_idx + 1);
       non_zero_idx < nn.size();
       non_zero_idx = indexNonZero(nn, non_zero_idx + 1)) {
    std::swap(nn[zero_idx], nn[non_zero_idx]);
    ++zero_idx;
  }
}

}  // namespace iq::shiftzeros
